app/infra/QueueAdapter.ts

```diff
@@ -0,0 +1,47 @@
+import {
+  AccessLevel,
+  Inject,
+  ContextProto,
+} from '@eggjs/tegg';
+import { Redis } from 'ioredis';
+import { QueueAdapter } from '../common/typing';
+
+/**
+ * Use sort set to keep queue in order and keep same value only insert once
+ */
+@ContextProto({
+  accessLevel: AccessLevel.PUBLIC,
+  name: 'queueAdapter',
+})
+export class RedisQueueAdapter implements QueueAdapter {
+  @Inject()
+  private readonly redis: Redis;
+
+  private getQueueName(key: string) {
+    return `CNPMCORE_Q_V2_${key}`;
+  }
+
+  private getQueueScoreName(key: string) {
+    return `CNPMCORE_Q_S_V2_${key}`;
+  }
+
+  /**
+   * If queue has the same item, return false
+   * If queue not has the same item, return true
+   */
+  async push<T>(key: string, item: T): Promise<boolean> {
+    const score = await this.redis.incr(this.getQueueScoreName(key));
```

Comment:
First, Redis is not a good choice for the task queue. I think it would be better to use MQ or other tools.

Second, maybe you can use use Lua script to keep the command atomic

Reply:
MQ is good for queue, but we choice less infrastructure dependency so use the redis.

Reply:
Yes, lua script is better, we should use it.

Reply:
And infra can be replace with other implement. It's ok to use mq impl the new queue adapter.